pallets/gear-program/src/migrations.rs

```diff
@@ -0,0 +1,342 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{Config, MemoryPages, Pallet, PausedProgramStorage, ProgramStorage, ResumeSessions};
+use common::Program;
+use frame_support::{
+    traits::{Get, GetStorageVersion, OnRuntimeUpgrade},
+    weights::Weight,
+};
+use frame_system::pallet_prelude::BlockNumberFor;
+use gear_core::program::MemoryInfix;
+use sp_std::marker::PhantomData;
+
+#[cfg(feature = "try-runtime")]
+use {
+    frame_support::{
+        codec::{Decode, Encode},
+        traits::StorageVersion,
+    },
+    sp_std::vec::Vec,
+};
+
+const MEMORY_INFIX: MemoryInfix = MemoryInfix::new(0);
+
+pub struct MigrateToV3<T: Config>(PhantomData<T>);
+
+impl<T: Config> OnRuntimeUpgrade for MigrateToV3<T> {
+    #[cfg(feature = "try-runtime")]
+    fn pre_upgrade() -> Result<Vec<u8>, &'static str> {
+        assert_eq!(
+            StorageVersion::get::<Pallet<T>>(),
+            2,
+            "Can only upgrade from version 2"
+        );
+
+        let count = v2::ProgramStorage::<T>::iter().fold(0u64, |count, (program_id, program)| {
+            match program {
+                v2::Program::Terminated(_) | v2::Program::Exited(_) => {
+                    assert!(v2::MemoryPageStorage::<T>::iter_key_prefix(program_id)
+                        .next()
+                        .is_none());
+                }
+                v2::Program::Active(_) => (),
+            }
+
+            count + 1
+        });
+
+        Ok(count.encode())
+    }
+
+    fn on_runtime_upgrade() -> Weight {
+        let current = Pallet::<T>::current_storage_version();
+        let onchain = Pallet::<T>::on_chain_storage_version();
+
+        log::info!(
+            "🚚 Running migration with current storage version {current:?} / onchain {onchain:?}"
+        );
+
+        // 1 read for on chain storage version
+        let mut weight = T::DbWeight::get().reads(1);
+
+        if current == 3 && onchain == 2 {
+            ProgramStorage::<T>::translate(
+                |program_id, program: v2::Program<BlockNumberFor<T>>| {
+                    weight = weight.saturating_add(T::DbWeight::get().reads_writes(1, 1));
+
+                    Some(match program {
+                        v2::Program::Active(p) => {
+                            for (page, data) in v2::MemoryPageStorage::<T>::drain_prefix(program_id)
+                            {
+                                weight =
+                                    weight.saturating_add(T::DbWeight::get().reads_writes(1, 1));
```

Comment:
One removal, one insert are 2 writes
```suggestion
                                    weight.saturating_add(T::DbWeight::get().reads_writes(0, 2));
```